### ISSUE TITLE
Make host_to_device and device_to_host functions more general.

### ISSUE THREAD

**Issue by bartgol** [E3SM-Project/EKAT#16]
The functions require a bunch of Kokkos::Arrays. However, since they must be called from host (since they call deep_copy), there's no need to use Kokkos::Array. Any container would do. To be extremist, one could use something like

```
template<typename ViewT>
void host_to_device (typename ViewT::value_type::scalar_type** data,
                     ViewT* views,
                     int dim1, int dim2)
```
but that is a tad too C-like, maybe. A flexible implementation could be
```
template<typename ViewArrayT,
         typename SizeArrayT,
         typename DataArrayT>
void host_to_device (DataArrayT& data, ViewArrayT& views, SizeArrayT& sizes)
```
Inside the impl, we would have some checks:
```
using ViewT = decltype(views[0]);
using PackT = typename ViewT::non_const_value_type;
using ScalarT = typename PackT::scalar_type;
using DataT = typename std::remove_cv<typename std::remove_pointer<decltype(data[0])>::type>::type;

static_assert (std::is_same<DataT,ScalarT>::value, "Error! Incompatible data types.");
EKAT_ASSERT_MSG(views.size()==data.size(), "Error! Incompatible input sizes.");
EKAT_ASSERT_MSG(views.size()==sizes.size(), "Error! Incompatible input sizes.");
// Inside the loop:
EKAT_ASSERT_MSG(view[i].size()==pack::npack<PackT>(sizes[i])`, "Error! Input view badly sized.").
```

The benefit of this impl is that we can use whatever array/vector implementation is easier, as long as they provide a `size()` method and overload `operator[]`.

Also, `host_to_device` should not allocate the views, in case the user already took notes of the stored ptrs. We may optionally allocate them if explicitly asked though, via an extra bool arg (defaulting to false).

**Comment by jeff-cohere** [comment]
What do you think about these alternative names for the template parameters?

* `ArrayViewT` instead of `ViewArrayT`
* `ArraySizeT` instead of `SizeArrayT`
* `ArrayDataT` instead of `DataArrayT`

**Comment by bartgol** [comment]
I think in English the first word is a specifier of the second. So `ArraySize` sounds like the size of an array, while `SizeArray` sounds like an array of sizes.

**Comment by jeff-cohere** [comment]
Oh, yeah, good point! I guess I misunderstood the forms!

**Comment by jgfouca** [comment]
This was fixed by #80 .
